src/WMBot/Commands/Admin.cs

```diff
@@ -109,6 +110,89 @@ private static void CommandList(CommandParams parameters)
                 IRC.DeliverMessage("I know: " + commands, parameters.SourceUser);
         }
 
+        private static void Restrict(CommandParams parameters)
+        {
+            if (string.IsNullOrEmpty(parameters.Parameters))
+            {
+                IRC.DeliverMessage("Please specify restrict action", parameters.SourceChannel);
+                return;
+            }
+
+            List<string> pm = new List<string>(parameters.Parameters.Trim().Split(' '));
+            switch (pm[0])
+            {
+                case "add":
+                    if (pm.Count == 1)
+                    {
+                        IRC.DeliverMessage("This command requires channel name to be specified", parameters.SourceChannel.Name);
+                        break;
+                    }
+
+                    string channel_name = pm[1];
+                    bool allow_adding_by_admin = pm.Count >= 3 && pm[2] == "admin";
+
+                    if (!Core.ValidFile(channel_name) || !channel_name.StartsWith("#"))
+                    {
+                        IRC.DeliverMessage(messages.Localize("InvalidName", parameters.SourceChannel.Language, new List<string> { channel_name }), parameters.SourceChannel);
+                        return;
+                    }
+
+                    parameters.SourceChannel.PrimaryInstance.Network.Transfer("PART " + channel_name);
+
+                    Channel channel = new Channel(channel_name) {RestrictedToJoin = true, AdminCanAddToRestrictedChannel = allow_adding_by_admin};
+                    lock (Configuration.Channels)
+                    {
+                        bool new_channel = true;
+                        foreach (Channel cu in Configuration.Channels)
+                        {
+                            if (channel_name == cu.Name)
+                            {
+                                new_channel = false;
```

Comment:
I see that you do the check here, and not insert the newly created channel in case there already is one. I would probably do it the other way and create new object only when I am sure it's necessary, but this would probably work as well.